src/concrete/ml/sklearn/base.py

```diff
@@ -1690,3 +1695,388 @@ def predict_proba(self, X: Data, fhe: Union[FheMode, str] = FheMode.DISABLE) ->
+            },
+        ).model
+
+        self._clean_graph()
+
+    def _clean_graph(self) -> None:
+        """Clean the ONNX graph from undesired nodes."""
+        assert self.onnx_model_ is not None, self._is_not_fitted_error_message()
+
+        # Remove cast operators as they are not needed
+        remove_node_types(onnx_model=self.onnx_model_, op_types_to_remove=["Cast"])
+
+    def fit(self, X: Data, y: Target, **fit_parameters):
+        # Reset for double fit
+        self._is_fitted = False
+
+        # KNeighbors handles multi-labels data
+        X, y = check_X_y_and_assert_multi_output(X, y)
+
+        self._y = y
+
+        # Fit the scikit-learn model
+        self._fit_sklearn_model(X, y, **fit_parameters)
+
+        # Check that the underlying sklearn model has been set and fit
+        assert self.sklearn_model is not None, self._sklearn_model_is_not_fitted_error_message()
+
+        # Retrieve the ONNX graph
+        self._set_onnx_model(X)
+
+        # Convert the n_bits attribute into a proper dictionary
+        n_bits = get_n_bits_dict(self.n_bits)
+
+        input_n_bits = n_bits["op_inputs"]
+        input_options = QuantizationOptions(n_bits=input_n_bits, is_signed=True)
+
+        # Quantize the inputs and store the associated quantizer
+        q_inputs = QuantizedArray(n_bits=input_n_bits, values=X, options=input_options)
+        input_quantizer = q_inputs.quantizer
+        self.input_quantizers.append(input_quantizer)
+
+        weights_n_bits = n_bits["op_weights"]
+        weight_options = QuantizationOptions(n_bits=weights_n_bits, is_signed=True)
+
+        # Quantize the _X_fit and store the associated quantizer
+        # Weights in KNN algorithms are the train data points
+        # pylint: disable-next=protected-access
+        _X_fit = self.sklearn_model._fit_X
+        q_X_fit = QuantizedArray(
+            n_bits=n_bits["op_weights"],
+            values=numpy.expand_dims(_X_fit, axis=1) if len(_X_fit.shape) == 1 else _X_fit,
+            options=weight_options,
+        )
+        self._q_X_fit = q_X_fit.qvalues
+        self._q_X_fit_quantizer = self._weight_quantizer = q_X_fit.quantizer
+
+        # mypy
+        assert self._q_X_fit_quantizer.scale is not None
+
+        # We assume that the query has the same distribution as the data in _X_fit.
+        # therefore, they use the same scaling and zero point.
+        # https://arxiv.org/abs/1712.05877
+
+        self.output_quant_params = UniformQuantizationParameters(
+            scale=self._q_X_fit_quantizer.scale,
+            zero_point=self._q_X_fit_quantizer.zero_point,
+            offset=0,
+        )
+
+        output_quantizer = UniformQuantizer(params=self.output_quant_params, no_clipping=True)
+
+        # Since the matmul and the bias both use the same scale and zero-points, we obtain that
+        # y = S*(q_y - 2*Z) when de-quantizing the values. We therefore need to multiply the initial
```

Comment:
this comment is not relevant anymoe, there is not bias term here !

Reply:
You are right, I forgot to remove it.
Good catch, thanks